### ISSUE TITLE
Feature: Child logger hook

### ISSUE THREAD

**Issue by brettwillis** [fastify/fastify#4751]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

As far as I can tell from Pino, we get a bit of a performance gain for child logger bindings that are provided when the child logger is instantiated (the bindings are pre-serialised).

Currently, Fastify just binds the request ID when creating a child logger:
https://github.com/fastify/fastify/blob/c4cb3a87cdbd33524556314341aa2b9bea61422f/lib/route.js#L399-L414

I propose to have a hook function (as part of the Fastify config) that will enable us to add bindings to the child logger, with access to the request object.

### Motivation

There are any number of scenarios where some logger bindings can be calculated on a per-request basis, other than just the request ID.

One such example is in Google Cloud Logging, where the trace context consists of **multiple** fields (not just one field), such as the `trace` and `spanId`, which should be parsed/calculated once *and only once* when the child logger is created, and *immediately when created* (not in a subsequent hook) so that all log entries (even early error logs if we don't even to process any hooks) get the bindings.

In such cases it would be great to have the pre-serialised performance win of child bindings, and also convenient access to the request object when binding.

### Example

Perhaps something like below would allow us to arbitrarily modify or add to the child bindings for a request's child logger before that logger is created.

```diff
function routeHandler (req, res, params, context, query) { 
   const id = genReqId(req) 
  
-  const loggerBinding = { 
+  let loggerBinding = { 
     [requestIdLogLabel]: id 
   } 

+  if (bindRequestLoggerFn) {
+    loggerBinding = bindRequestLoggerFn(loggerBinding, req) || {};
+  }

   const loggerOpts = { 
     level: context.logLevel 
   } 
  
   if (context.logSerializers) { 
     loggerOpts.serializers = context.logSerializers 
   } 
   const childLogger = logger.child(loggerBinding, loggerOpts) 
   childLogger[kDisableRequestLogging] = disableRequestLogging
```

**Comment by mcollina** [comment]
This could be a nice option for the Fastify constructor.

**Comment by brettwillis** [comment]
@mcollina Yep I'll work on a PR. In my mind, I'm debating whether this belongs as a global configuration (on the Fastify constructor logger options), or as an encapsulated hook (like `onRequest`) which can be encapsulated on a per-plugin (or per-route) basis.

For the use case I personally have in mind, a global configuration will be fine because it depends only on the platform that the app runs on. However I guess it is conceivable that different plugins/routes may receive requests from different platforms which may need to parse bindings in different ways? Or maybe that's overcomplicating it.

Thoughts?

**Comment by metcoder95** [comment]
>For the use case I personally have in mind, a global configuration will be fine because it depends only on the platform that the app runs on. However, I guess it is conceivable that different plugins/routes may receive requests from different platforms which may need to parse bindings in different ways? Or maybe that's overcomplicating it.

The `fastify` constructor seems enough, as anyways it passes the `request` object that can be used to retrieve the data out of it to create the child logger. On top, maybe an API like `fastify.setChildLoggerFactory` (the name is just a suggestion, might not be best 😅) can help this be used as well by a `plugin` if any. 

Not so sure if it really deserves its own hook 🤔

**Comment by brettwillis** [comment]
Ok sounds fair to me. Your comment raised another question: should this be

1. a hook to modify the bindings object before being passed to `logger.child()` internally (this is what I initially thought) e.g.
```Javascript
fastify({
  ...
  bindChildLogger: (bindings, loggerOpts, req) => {
    bindings.myKey = ... // calculate something base on req
    bindings.reqId // do something to the base bindings (request ID)
  }
})
```
2. a hook to create the child logger itself (e.g. take the base bindings and return a new logger instance) as alluded by your "logger factory" name
```Javascript
fastify({
  ...
  childLoggerFactory: (logger, bindings, loggerOpts, req) => {
    bindings.myKey = ... // calculate something base on req
    return logger.child(bindings, loggerOpts)
  }
})

We are already able to pass our own logger implementation anyway (so we can have our own `logger.child()` implementation), however we do not have access to the request object from there. The second option is untimately more flexible I guess.

**Comment by brettwillis** [comment]
@metcoder95 @mcollina before I go too far on the PR, do you have a preference between options 1/2 above?

**Comment by metcoder95** [comment]
>@metcoder95 @mcollina before I go too far on the PR, do you have a preference between options 1/2 above?

Option 2 seems to offer the best flexibility and ergonomics. In that way, the child logger instantiation can be customised at full if required 👍
